src/data/list/basic.lean

```diff
@@ -490,6 +490,20 @@ theorem concat_cons (a b : α) (l : list α) : concat (a :: l) b = a :: concat l
 @[simp] theorem concat_eq_append (a : α) (l : list α) : concat l a = l ++ [a] :=
 by induction l; simp only [*, concat]; split; refl
 
+theorem init_eq_of_concat_eq {a : α} {l₁ l₂ : list α} : concat l₁ a = concat l₂ a → l₁ = l₂ :=
```

Comment:
Not sure about the name, but maybe:
```suggestion
theorem eq_of_concat_eq {a : α} {l₁ l₂ : list α} : concat l₁ a = concat l₂ a → l₁ = l₂ :=
```
Since `init` isn't the name of any function used in the statement, I would avoid using it in the name of the declaration.

Reply:
On second thought, your original names might be perfectly OK. Second opinion, anyone?

Reply:
If it helps, `head_eq_of_cons_eq` and `tail_eq_of_cons_eq` don't mention `head` or `tail` either 😛

Reply:
OK, I'm convinced. :)